algo/src/test/kotlin/com/hartwig/actin/algo/evaluation/treatment/HasHadTreatmentResponseFollowingSomeTreatmentOrCategoryOfTypesTest.kt

```diff
@@ -0,0 +1,325 @@
+    }
+
+    @Test
+    fun `Should fail if treatment history does not contain target treatment`() {
+        val history = listOf(
+            TreatmentTestFactory.treatmentHistoryEntry(setOf(WRONG_SPECIFIC_TREATMENT)),
+            TreatmentTestFactory.treatmentHistoryEntry(setOf(WRONG_CATEGORY_TREATMENT))
+        )
+        val wrongTypeHistory = listOf(TreatmentTestFactory.treatmentHistoryEntry(setOf(WRONG_TYPE_TREATMENT)))
+        assertForAllClinicalBenefitFunctions(EvaluationResult.FAIL, TreatmentTestFactory.withTreatmentHistory(history))
+        assertForAllCompleteResponseFunctions(EvaluationResult.FAIL, TreatmentTestFactory.withTreatmentHistory(history))
+        assertEvaluation(
+            EvaluationResult.FAIL,
+            functionWithSpecificCategoryAndTypeAndClinicalBenefit.evaluate(TreatmentTestFactory.withTreatmentHistory(wrongTypeHistory))
+        )
+        assertEvaluation(
+            EvaluationResult.FAIL,
+            functionWithSpecificCategoryAndTypeAndCompleteResponse.evaluate(TreatmentTestFactory.withTreatmentHistory(wrongTypeHistory))
+        )
+    }
+
+    @Test
+    fun `Should pass if treatment history contains target treatment with best response complete response`() {
+        listOf(CORRECT_TREATMENT, OTHER_CORRECT_TREATMENT).forEach {
+            val history = listOf(
+                TreatmentTestFactory.treatmentHistoryEntry(
+                    setOf(it),
+                    bestResponse = TreatmentResponse.COMPLETE_RESPONSE
+                )
+            )
+            assertForAllClinicalBenefitFunctions(EvaluationResult.PASS, TreatmentTestFactory.withTreatmentHistory(history))
+            assertForAllCompleteResponseFunctions(EvaluationResult.PASS, TreatmentTestFactory.withTreatmentHistory(history))
+        }
+    }
+
+    @Test
+    fun `Should pass if treatment history contains target therapy combined with other therapy with best response complete response`() {
+        val history1 = listOf(
+            TreatmentTestFactory.treatmentHistoryEntry(
+                CORRECT_TREATMENT_WITH_OTHER_CATEGORY_COMBINATION,
+                bestResponse = TreatmentResponse.COMPLETE_RESPONSE
+            )
+        )
+        assertForAllClinicalBenefitFunctions(EvaluationResult.PASS, TreatmentTestFactory.withTreatmentHistory(history1))
+        assertForAllCompleteResponseFunctions(EvaluationResult.PASS, TreatmentTestFactory.withTreatmentHistory(history1))
+
+        val history2 = listOf(
+            TreatmentTestFactory.treatmentHistoryEntry(
+                TARGET_TREATMENT_WITH_SAME_CATEGORY_COMBINATION,
+                bestResponse = TreatmentResponse.COMPLETE_RESPONSE
+            )
+        )
+        assertForAllClinicalBenefitFunctions(EvaluationResult.PASS, TreatmentTestFactory.withTreatmentHistory(history2))
+        assertForAllCompleteResponseFunctions(EvaluationResult.PASS, TreatmentTestFactory.withTreatmentHistory(history2))
```

Comment:
I think it would be cleaner to factor out the treatments:
```suggestion
        listOf(CORRECT_TREATMENT_WITH_OTHER_CATEGORY_COMBINATION, TARGET_TREATMENT_WITH_SAME_CATEGORY_COMBINATION)
            .forEach { treatment ->
                val history = listOf(
                    TreatmentTestFactory.treatmentHistoryEntry(
                        treatment,
                        bestResponse = TreatmentResponse.COMPLETE_RESPONSE
                    )
                )
                assertForAllClinicalBenefitFunctions(EvaluationResult.PASS, TreatmentTestFactory.withTreatmentHistory(history))
                assertForAllCompleteResponseFunctions(EvaluationResult.PASS, TreatmentTestFactory.withTreatmentHistory(history))
            }
```